pkg/plugin/velero/item_snapshotter/v1/item_snapshotter.go

```diff
@@ -0,0 +1,139 @@
+/*
+Copyright the Velero contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"context"
+	"time"
+
+	"github.com/vmware-tanzu/velero/pkg/plugin/velero"
+
+	"k8s.io/apimachinery/pkg/runtime"
+
+	api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+)
+
+type AlsoHandlesInput struct {
+	ItemToSnapshot runtime.Unstructured
+	Backup         *api.Backup
+}
+
+type SnapshotItemInput struct {
+	Item   runtime.Unstructured
+	Params map[string]string
+	Backup *api.Backup
+}
+
+type SnapshotItemOutput struct {
+	UpdatedItem      runtime.Unstructured
+	SnapshotID       string
+	SnapshotMetadata map[string]string			// SnapshotMetadata is information in addition to the SnapshotID that the
+												// plugin wants to store in the backup.  SnapshotMetadata will be passed
+												// back in for DeleteSnapshot and CreateItemFromSnapshot
+	AdditionalItems  []velero.ResourceIdentifier
+	HandledItems     []velero.ResourceIdentifier // Items that were handled by this snapshot that can be ignored
+}
+
+// ProgressInput contains the input parameters for the ItemSnapshotter's Progress function.
+type ProgressInput struct {
+	// ItemID is the id of item that was stored in the backup
+	ItemID velero.ResourceIdentifier
```

Comment:
I notice here's a little inconsistency, in `SnapshotItemInput` the attribute is `Item` but in `ProgressInput` the attribute is `ItemID`

Reply:
When we snapshot we pass in the whole resource in case anything in it needs to be used.  On progress we only need the ID and we may only have the ID at that point.  If the Velero server is restarted before the progress is finished, we will lose the in-memory copy.  The resource may be modified or deleted after the backup finishes so in order to pass in the resource here we would need to extract it from the backup.  There's no need for anything other than the ID for checking progress so this saves a lot of trouble trying to recover the whole resource.